src/topology/homotopy/homotopy_group.lean

```diff
@@ -154,42 +314,140 @@ begin
-{ to_fun := λ p, path.mk ⟨λ t, p (λ _, t), by {continuity, exact p.1.2}⟩
-    (p.boundary (λ _, 0) ⟨0, or.inl rfl⟩)
-    (p.boundary (λ _, 1) ⟨1, or.inr rfl⟩),
+/-- The 1-dimensional generalized loops based at `x` are in 1-1 correspondence with loops at `x`. -/
+@[simps] def gen_loop_one_equiv_loop : Ω^(fin 1) x ≃ Ω X x :=
+{ to_fun := λ p, path.mk ⟨λ t, p (λ _, t), by continuity⟩
+    (gen_loop.boundary _ (λ _, 0) ⟨0, or.inl rfl⟩)
+    (gen_loop.boundary _ (λ _, 1) ⟨1, or.inr rfl⟩),
   inv_fun := λ p,
-  { to_fun := λ c, p c.head,
-    boundary := begin
-      rintro y ⟨i, iH|iH⟩; cases unique.eq_default i;
-      apply (congr_arg p iH).trans, exacts [p.source, p.target],
-    end },
-  left_inv := λ p, by { ext1, exact congr_arg p y.one_char.symm },
+  begin
+    refine ⟨⟨λ (c : I^fin 1), p (cube.head c), by continuity⟩, _⟩,
+    rintro y ⟨i, iH|iH⟩; cases unique.eq_default i;
+    apply (congr_arg p iH).trans, exacts [p.source, p.target],
+  end,
+  left_inv := λ p, by { ext, exact congr_arg p (cube.one_char y).symm },
   right_inv := λ p, by { ext, refl } }
 
-/--
-The first homotopy group at `x` is equivalent to the fundamental group,
-i.e. the loops based at `x` up to homotopy.
--/
-def pi1_equiv_fundamental_group : π 1 x ≃ fundamental_group X x :=
+/-- The first homotopy group at `x` is equivalent to the fundamental group, i.e. the loops based at
+  `x` up to homotopy. -/
+-- TODO: deduce from homotopy_group_equiv_fundamental_group?
+def pi1_equiv_fundamental_group : π_ 1 X x ≃ fundamental_group X x :=
 begin
   refine equiv.trans _ (category_theory.groupoid.iso_equiv_hom _ _).symm,
-  refine quotient.congr gen_loop_one_equiv_path_self _,
-  -- homotopic iff homotopic
+  refine quotient.congr gen_loop_one_equiv_loop _,
   intros, split; rintros ⟨H⟩,
   exacts
   [⟨{ to_fun := λ tx, H (tx.fst, λ _, tx.snd),
       map_zero_left' := λ _, by convert H.apply_zero _,
       map_one_left' := λ _, by convert H.apply_one _,
-      prop' := λ t y iH, H.prop' _ _ ⟨0,iH⟩ }⟩,
-   ⟨{ to_fun := λ tx, H (tx.fst, tx.snd.head),
-      map_zero_left' := λ y, by { convert H.apply_zero _, exact y.one_char },
-      map_one_left' := λ y, by { convert H.apply_one _, exact y.one_char },
+      prop' := λ t y iH, H.prop' _ _ ⟨0, iH⟩ }⟩,
+   ⟨{ to_fun := λ tx, H (tx.fst, (cube.head tx.snd)),
+      map_zero_left' := λ y, by { convert H.apply_zero _, exact cube.one_char y },
+      map_one_left' := λ y, by { convert H.apply_one _, exact cube.one_char y },
       prop' := λ t y ⟨i, iH⟩, begin
         cases unique.eq_default i, split,
-        { convert H.eq_fst _ _, exacts [y.one_char, iH] },
-        { convert H.eq_snd _ _, exacts [y.one_char, iH] },
+        { convert H.eq_fst _ _, exacts [cube.one_char y, iH] },
+        { convert H.eq_snd _ _, exacts [cube.one_char y, iH] },
       end }⟩],
 end
+
+section
+variables {n : ℕ} (i : N)
+
+/-- Group structure on `π_(n+1)`. -/
+instance group : group (π_(n+1) X x) :=
```

Comment:
Is there a reason this is not an instance on `homotopy_group`?

EDIT: Oh, `N` doesn't have a canonical element. Ignore this comment.

Reply:
This should be on `π_n` assuming `[ne_zero n]` 

It should probably also exist for any type `N` with element `n`, but that requires a new variant of `homotopy_group` that takes `n` as (unused) argument, so this should probably wait for a later PR.

Reply:
We could use `[inhabited N]` here. Or even `[nonempty N]` and use choice, since the end result doesn't depend on the choice by `mul_spec`.

Reply:
Yeah, that is probably useful. And indeed, the use of `nonempty` seems reasonable here.

Reply:
Generalized to arbitrary `nonempty` type.